File: dbiz_app/dbiz_app/dbiz_app/report/inv01___bảng_kê_phiếu_nhập_kho_mua_hàng/inv01___bảng_kê_phiếu_nhập_kho_mua_hàng.js
// Copyright (c) 2024, lamnl and contributors
// For license information, please see license.txt

frappe.query_reports["INV01 - Bảng kê phiếu nhập kho mua hàng"] = {
	filters: [
		{
			fieldname: "start_date",
			label: __("From Date"),
			fieldtype: "Date",
			default: frappe.datetime.add_months(frappe.datetime.now_date(), -1),
			reqd: 1,
		},
		{
			fieldname: "end_date",
			label: __("To Date"),
			fieldtype: "Date",
			default: frappe.datetime.now_date(),
			reqd: 1,
		},
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			default: frappe.defaults.get_user_default("Company"),
			reqd: 1,
			on_change: function () {
				const company = frappe.query_report.get_filter_value("company");

				frappe.db.get_value("Company", company, ["company_name", "tax_id"], (r) => {
					frappe.query_report.set_filter_value("company_name", r.company_name);
					frappe.query_report.set_filter_value("tax_id", r.tax_id);
				});

				// Get Address Linked to the Company
				frappe.call({
					method: "dbiz_app.api.get_company_address",
					args: {
						company: company
					},
					callback: function (r) {
						if (r.message) {
							frappe.query_report.set_filter_value("company_address", r.message);
						}
					}
				});
			},
		},
		{
			fieldname: "company_name",
			fieldtype: "Data",
			hidden: 1, // Hidden as it is used only for display in the HTML when printing
		},
		{
			fieldname: "company_address",
			fieldtype: "Data",
			hidden: 1, // Hidden as it is used only for display in the HTML when printing
		},
		{
			fieldname: "tax_id",
			fieldtype: "Data",
			hidden: 1, // Hidden as it is used only for display in the HTML when printing
		},
		{
			fieldname: "warehouse",
			label: __("Warehouse"),
			fieldtype: "Link",
			options: "Warehouse",
			reqd: 0,
			get_query: () => ({
				filters: {
					"company": frappe.query_report.get_filter_value("company"),
					"is_group": 0,
				}
			}),
			on_change: function () {
				const warehouse = frappe.query_report.get_filter_value("warehouse");

				if (warehouse) {
					frappe.db.get_value("Warehouse", warehouse, "warehouse_name", (r) => {
						frappe.query_report.set_filter_value("warehouse_name", r.warehouse_name || "");
					});
				}
			},
		},
		{
			fieldname: "warehouse_name",
			fieldtype: "Data",
			hidden: 1, // Hidden as it is used only for display in the HTML when printing
		},
		{
			fieldname: "supplier",
			label: __("Supplier"),
			fieldtype: "Link",
			options: "Supplier",
			reqd: 0,
			on_change: function () {
				const supplier = frappe.query_report.get_filter_value("supplier");

				if (supplier) {
					frappe.db.get_value("Supplier", supplier, "supplier_name", (r) => {
						frappe.query_report.set_filter_value("supplier_name", r.supplier_name || "");
					});
				}
			},
		},
		{
			fieldtype: "Data",
			fieldname: "supplier_name",
			fieldtype: "Data",
			hidden: 1, // Hidden as it is used only for display in the HTML when printing
		},
		{
			fieldname: "item_group",
			label: __("Item Group"),
			fieldtype: "Link",
			options: "Item Group",
			reqd: 0,
			get_query: () => ({
				filters: {
					"is_group": 0,
				}
			}),
			on_change: function () {
				const item_group = frappe.query_report.get_filter_value("item_group");

				if (item_group) {
					frappe.db.get_value("Item Group", item_group, "item_group_name", (r) => {
						frappe.query_report.set_filter_value("item_group_name", r.item_group_name || "");
					});
				}
			},
		},
		{
			fieldname: "item_group_name",
			fieldtype: "Data",
			hidden: 1, // Hidden as it is used only for display in the HTML when printing
		},
		{
			fieldname: "account",
			label: __("Account"),
			fieldtype: "Link",
			options: "Account",
		},
	],
	// Trigger on_change manually after the report loads
	onload: function (query_report) {
		const company = frappe.query_report.get_filter_value("company");

		if (company) {
			query_report.get_filter("company").on_change(query_report);
		}

		window.format_number_js = (
			value,
			decimals = 0,
			number_format = frappe.boot.sysdefaults.number_format || "#.###,##"
		) => {
			const decimal_places =
				decimals === "default"
					? frappe.boot.sysdefaults.float_precision
					: decimals;
			return !isNaN(value)
				? format_number(parseFloat(value), number_format, decimal_places)
				: "";
		};
	}
};

File: dbiz_app/dbiz_app/dbiz_app/report/inv01___bảng_kê_phiếu_nhập_kho_mua_hàng/inv01___bảng_kê_phiếu_nhập_kho_mua_hàng.py
# Copyright (c) 2024, lamnl and contributors
# For license information, please see license.txt

import frappe
from frappe import _


def execute(filters=None):
    if not filters:
        filters = {}

    # Validate the dates before proceeding
    validate_dates(filters)

    # Define columns
    columns = get_columns()

    # SQL query with parameters from filters
    data = get_data(filters)

    return columns, data


def get_columns():
    return [
        {
            "label": "Mã kho",
            "fieldname": "warehouse_code",
            "fieldtype": "Data",
        },
        {
            "label": "Tên kho",
            "fieldname": "warehouse_name",
            "fieldtype": "Data",
        },
        {
            "label": "Loại chứng từ",
            "fieldname": "ct_type",
            "fieldtype": "Data",
        },
        {
            "label": "Ngày",
            "fieldname": "posting_date",
            "fieldtype": "Data",
        },
        {
            "label": "Số CT nhập kho",
            "fieldname": "ct_code",
            "fieldtype": "Data",
        },
        {
            "label": "Số đơn hàng/hoá đơn",
            "fieldname": "order_no",
            "fieldtype": "Data",
        },
        {
            "label": "Nhà cung cấp",
            "fieldname": "supplier_name",
            "fieldtype": "Data",
        },
        {
            "label": "Trạng thái",
            "fieldname": "status",
            "fieldtype": "Data",
        },
        {
            "label": "Diễn giải",
            "fieldname": "description",
            "fieldtype": "Data",
        },
        {
            "label": "Nhóm sản phẩm",
            "fieldname": "item_group",
            "fieldtype": "Data",
        },
        {
            "label": "Mã sản phẩm",
            "fieldname": "item_code",
            "fieldtype": "Data",
        },
        {
            "label": "Tên sản phẩm",
            "fieldname": "item_name",
            "fieldtype": "Data",
        },
        {
            "label": "ĐVT",
            "fieldname": "stock_uom",
            "fieldtype": "Data",
        },
        {
            "label": "Giá",
            "fieldname": "rate",
            "fieldtype": "Float",
        },
        {
            "label": "SL",
            "fieldname": "stock_qty",
            "fieldtype": "Float",
        },
        {
            "label": "Thành tiền",
            "fieldname": "total_amount",
            "fieldtype": "Float",
        },
    ]


def get_data(filters):
    conditions = get_conditions(filters)
    data = frappe.db.sql(
        """
        SELECT 
            tpri.warehouse AS warehouse_code,
            tw.warehouse_name,
            'Nhập mua hàng' AS ct_type,
            DATE_FORMAT(tpr.posting_date, '%%d/%%m/%%Y') AS posting_date,
            tpr.name AS ct_code,
            tpri.purchase_order AS order_no,
            tpr.supplier_name,
            tpr.status,
            tpri.description,
            tpri.item_group,
            tpri.item_code,
            tpri.item_name,
            tpri.stock_uom,
            tpri.rate,
            ABS(tpri.stock_qty) AS stock_qty,
            tpri.rate * ABS(tpri.stock_qty) AS total_amount
        FROM
            `tabPurchase Receipt Item` tpri
        LEFT JOIN `tabPurchase Receipt` tpr ON
            tpr.name = tpri.parent
        LEFT JOIN `tabWarehouse` tw ON
            tw.name = tpri.warehouse
        LEFT JOIN `tabPurchase Order` tpo ON
            tpo.name = tpri.purchase_order
        WHERE
            tpr.posting_date BETWEEN DATE(%(start_date)s) AND DATE(%(end_date)s)
            AND tpr.company = %(company)s
            AND (tpri.warehouse = %(warehouse)s OR %(warehouse)s IS NULL OR %(warehouse)s = '')
            AND (tpr.supplier = %(supplier)s OR %(supplier)s IS NULL OR %(supplier)s = '')
            AND (tpri.item_group = %(item_group)s OR %(item_group)s IS NULL OR %(item_group)s = '')
            AND (tw.account = %(account)s OR %(account)s IS NULL OR %(account)s = '')
        """,
        conditions,  # filters
        as_dict=True,
    )
    return data


def get_conditions(filters):
    conditions = {}
    for key, value in filters.items():
        if filters.get(key):
            conditions[key] = value
        if not filters.get("warehouse"):
            conditions["warehouse"] = None
        if not filters.get("supplier"):
            conditions["supplier"] = None
        if not filters.get("item_group"):
            conditions["item_group"] = None
        if not filters.get("account"):
            conditions["account"] = None
    return conditions


def validate_dates(filters):
    if filters.start_date > filters.end_date:
        frappe.throw(_("From Date must be before To Date"))
